File: LeetCodeCPP/345.reverse-vowels-of-a-string.cpp
/*
 * @lc app=leetcode id=345 lang=cpp
 *
 * [345] Reverse Vowels of a String
 */

#include <string>
#include <unordered_set>

using namespace std;

// @lc code=start
class Solution
{
public:
    string reverseVowels(string s)
    {
        int i = 0, j = s.size() - 1;
        while (i < j)
        {
            i = s.find_first_of("aeiouAEIOU", i);
            j = s.find_last_of("aeiouAEIOU", j);
            if (i < j)
            {
                swap(s[i++], s[j--]);
            }
        }
        return s;
    }
};
// @lc code=end

File: LeetCodePython/345.reverse-vowels-of-a-string.py
#
# @lc app=leetcode id=345 lang=python3
#
# [345] Reverse Vowels of a String
#


# @lc code=start
class Solution:
    def reverseVowels(self, s: str) -> str:
        characters = list(s)
        vowels = set("aeiouAEIOU")
        i = 0
        j = len(s) - 1
        while i < j:
            if characters[i] not in vowels:
                i += 1
            elif characters[j] not in vowels:
                j -= 1
            else:
                characters[i], characters[j] = characters[j], characters[i]
                i += 1
                j -= 1

        return "".join(characters)


# @lc code=end
